sw/device/tests/alert_handler_lpg_reset_toggle.c

```diff
@@ -0,0 +1,548 @@
+      .local_alerts_len = ARRAYSIZE(loc_alerts),
+      .classes = classes,
+      .class_configs = class_configs,
+      .classes_len = ARRAYSIZE(class_configs),
+      .ping_timeout = ping_timeout,
+  };
+
+  alert_handler_testutils_configure_all(&alert_handler, config,
+                                        kDifToggleEnabled);
+  // Enables alert handler irq.
+  CHECK_DIF_OK(dif_alert_handler_irq_set_enabled(
+      &alert_handler, kDifAlertHandlerIrqClassa, kDifToggleEnabled));
+
+  CHECK_DIF_OK(dif_alert_handler_irq_set_enabled(
+      &alert_handler, kDifAlertHandlerIrqClassb, kDifToggleEnabled));
+}
+
+/**
+ * A utility function to wait enough until the alert handler pings a peripheral
+ * alert
+ */
+void wait_enough_for_alert_ping() {
+  // wait enough
+  if (kDeviceType == kDeviceFpgaCw310) {
+    // NUM_ALERTS*2*margin_of_safety*(2**DW)*(1/kClockFreqPeripheralHz)
+    // (2**6)*2*4*(2**16)*(400ns) = 12.8s
+    busy_spin_micros(1000 * 12800);
+  } else if (kDeviceType == kDeviceSimDV) {
+    // NUM_ALERTS*2*margin_of_safety*(2**DW)*(1/kClockFreqPeripheralHz)
+    // (2**6)*2*4*(2**3)*(40ns) = 160us
+    busy_spin_micros(160);
+  } else {
+    // Verilator
+    // NUM_ALERTS*2*margin_of_safety*(2**DW)*(1/kClockFreqPeripheralHz)
+    // (2**6)*2*4*(2**16)*(8us) = 256s
+    // This seems to be impractical for the current clock frequency config
+    // of the Verilator tests (kClockFreqPeripheralHz = 125K).
+    LOG_FATAL("SUPPORTED PLATFORMS: DV and FPGA");
+    LOG_FATAL("TO SUPPORT THE PLATFORM %d, COMPUTE THE RIGHT WAIT-TIME",
+              kDeviceType);
+  }
+}
+
+/**
+ * Get `num` distinct random numbers in the range [0, `max`] from
+ * RV_CORE_IBEX_RND_DATA.
+ *
+ * @param ibex The Ibex DIF object.
+ * @param num The number of random numbers to get.
+ * @param[out] rnd_buf Pointer to the buffer to write the random numbers to.
+ * @param max The maximum random value returned.
+ */
+static void get_rand_words(dif_rv_core_ibex_t *ibex, int num, uint32_t *rnd_buf,
+                           uint32_t max) {
+  uint32_t rnd_word;
+  for (int i = 0; i < num; ++i) {
+    bool found = false;
+    while (found == false) {
+      // Get a new random number.
+      CHECK_DIF_OK(dif_rv_core_ibex_read_rnd_data(ibex, &rnd_word));
+      rnd_word = rnd_word % max;
+      // Check if the number is unique.
+      found = true;
+      for (int j = 0; j < i; ++j) {
+        if (rnd_buf[j] == rnd_word) {
+          // Start over.
+          found = false;
+          break;
+        }
+      }
+    }
+    // Add the number to the buffer.
+    rnd_buf[i] = rnd_word;
+  }
+}
+
+enum {
+  // Non-volatile counter for the test steps on FPGA.
+  kCounterTestSteps = 0,
+};
+
+/**
+ * External ISR.
+ *
+ * Handles all peripheral interrupts on Ibex. PLIC asserts an external interrupt
+ * line to the CPU, which results in a call to this OTTF ISR. This ISR
+ * overrides the default OTTF implementation.
+ */
+void ottf_external_isr(void) {
```

Comment:
Please correct me if I am wrong here - I think in this test we do not expect any interrupts from alert_handler.
Shall we just report error if interrupt ever fired?

Reply:
> Please correct me if I am wrong here - I think in this test we do not expect any interrupts from alert_handler. Shall we just report error if interrupt ever fired?

True, we don't expect any interrupts to be fired. I think, in the current code, the interrupts are silently acknowledged. I can put an always-failing CHECK macro (`CHECK(0,...)`) inside this function. It would cause the test fail in case any interrupt is fired. would that work?

Reply:
Yeah I think that would be good to do.

We did something similar in the NMI test as well, although in that case we checked upon test exit:
https://github.com/lowRISC/opentitan/blob/b0cdccec0f46da022ecc45cc2576cc9b9d785726/sw/device/tests/rv_core_ibex_nmi_irq_test.c#L291-L292

Both approaches are probably fine.

Reply:
thanks! I added a CHECK statement inside the interrupt handler, which will terminate the test in case an interrupt is triggered.